Enable load nested hint files

lifeizhou-ap opened the pull request:
Implement the feature for https://github.com/block/goose/issues/3724

**What**
- Added the configuration for users to choose whether to load the nested goose hint file
- When configuration is enabled, it will load the hints file up to the directory with .git dir or current directory if no .git found

**Next (not in this PR)**
It would be good to show loaded hints in CLI and Desktop.

alexhancock commented:
Tried it locally and asked:

"What .goosehints do you already know about?" from cwd of `goose/ui/desktop`

it picked up the one from `goose/ui/desktop/.goosehints` but not `goose/.goosehints`

I wonder if this should work, or indicates an issue with the upward traversal

michaelneale commented:
I think consider default on for this, if it isn't ready for that, then it needs more work IMO. This would use the nearest hints - is that right? or does it keep loading them all the way up?

alexhancock commented:
Yeah I agree with @michaelneale above! I think this would be good just "default to on" behavior.

I also don't know that it needs to be configureable tbh. The more options we add to settings the more conceptual overhead there is for the user. People could configure it by modifying/adjusting/removing the content of the goosehints files it finds and uses.

lifeizhou-ap commented:
> I think consider default on for this, if it isn't ready for that, then it needs more work IMO. 
👍 

> This would use the nearest hints - is that right? or does it keep loading them all the way up?
It will keep loading them all the way up until they see the directory with .git (eg project root). Otherwise, it uses the `.goosehints` in the current directory

lifeizhou-ap commented:
> Yeah I agree with @michaelneale above! I think this would be good just "default to on" behavior.
> 
> I also don't know that it needs to be configureable tbh. The more options we add to settings the more conceptual overhead there is for the user. People could configure it by modifying/adjusting/removing the content of the goosehints files it finds and uses.

👍 This totally makes sense!

lifeizhou-ap commented:
> Tried it locally and asked:
> 
> "What .goosehints do you already know about?" from cwd of `goose/ui/desktop`
> 
> it picked up the one from `goose/ui/desktop/.goosehints` but not `goose/.goosehints`
> 
> I wonder if this should work, or indicates an issue with the upward traversal

Hi @alexhancock Thank you so much for testing this! Really appreciate it!

I tried on local, it seems Goose picked up the `goose/.goosehints`.  I guess it might need to run the developer mcp to load the hints. (this is the existing behaviour). 

<img width="1313" height="971" alt="Screenshot 2025-08-22 at 12 21 21 pm" src="https://github.com/user-attachments/assets/b40a36b9-6bca-437d-ae3f-ebd71773b9d3" />

lifeizhou-ap commented:
Hi @alexhancock @michaelneale,

Thank you so much for your time to review this PR. I've replied to your comments and removed the NESTED_GOOSE_HINTS configuration as suggested. 

I added a bit clarification on MAX_DEPTH,  happy to increase if it is still helpful. 

Thank you!
